File: Problems 1000-1099/1061. Lexicograpically Smallest Equivalent String/solution.py
class UnionFind:
    def __init__(self, n):
        self.parent = [i for i in range(n)]
    
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    
    def union(self, x, y):
        rootX = self.find(x)
        rootY = self.find(y)

        if rootX == rootY:
            return
        
        if rootX < rootY:
            self.parent[rootY] = rootX
        else:
            self.parent[rootX] = rootY

class Solution:
    def smallestEquivalentString(self, s1: str, s2: str, baseStr: str) -> str:
        n = len(s1)
        uf = UnionFind(26)
        for i in range(n):
            uf.union(ord(s1[i])-ord('a'), ord(s2[i])-ord('a'))
        
        res = []
        for c in baseStr:
            i = uf.find(ord(c)-ord('a'))
            res.append(chr(i+ord('a')))
        
        return "".join(res)

File: Problems 1000-1099/1061. Lexicograpically Smallest Equivalent String/solution.cpp
class UnionFind {
    private:
        vector<int> parent;
    public:
        UnionFind(int n): parent(n) {
            for (int i = 0; i < n; i++) {
                parent[i] = i;
            }
        }
    
        int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }
    
        void unionByLex(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
    
            if (rootX == rootY) return;
    
            if (rootX < rootY) {
                parent[rootY] = rootX;
            } else {
                parent[rootX] = rootY;
            }
        }
    };
    
    class Solution {
    public:
        string smallestEquivalentString(string s1, string s2, string baseStr) {
            int n = s1.length();
            UnionFind uf(26);
            for (int i = 0; i < n; i++) {
                uf.unionByLex(s1[i]-'a', s2[i]-'a');
            }
    
            string res;
            for (char c : baseStr) {
                int i = uf.find(c-'a');
                res += i+'a';
            }
            return res;
        }
    };